openpype/hosts/maya/plugins/publish/extract_xgen.py

```diff
@@ -0,0 +1,142 @@
+import os
+import copy
+import tempfile
+
+from maya import cmds
+import xgenm
+
+from openpype.pipeline import publish
+from openpype.hosts.maya.api.lib import (
+    maintained_selection, attribute_values, write_xgen_file, delete_after
+)
+from openpype.lib import StringTemplate
+
+
+class ExtractXgenCache(publish.Extractor):
+    """Extract Xgen
+
+    Workflow:
+    - Duplicate nodes used for patches.
+    - Export palette and import onto duplicate nodes.
+    - Export/Publish duplicate nodes and palette.
+    - Export duplicate palette to .xgen file and add to publish.
+    - Publish all xgen files as resources.
+    """
+
+    label = "Extract Xgen"
+    hosts = ["maya"]
+    families = ["xgen"]
+    scene_type = "ma"
+
+    def process(self, instance):
+        if "representations" not in instance.data:
+            instance.data["representations"] = []
+
+        staging_dir = self.staging_dir(instance)
+        maya_filename = "{}.{}".format(instance.data["name"], self.scene_type)
+        maya_filepath = os.path.join(staging_dir, maya_filename)
+
+        # Get published xgen file name.
+        template_data = copy.deepcopy(instance.data["anatomyData"])
+        template_data.update({"ext": "xgen"})
+        templates = instance.context.data["anatomy"].templates["publish"]
+        xgen_filename = StringTemplate(templates["file"]).format(template_data)
+
+        xgen_path = os.path.join(
+            self.staging_dir(instance), xgen_filename
+        ).replace("\\", "/")
+        type = "mayaAscii" if self.scene_type == "ma" else "mayaBinary"
+
+        # Duplicate xgen setup.
+        with delete_after() as delete_bin:
+            duplicate_nodes = []
+            # Collect nodes to export.
+            for _, connections in instance.data["xgenConnections"].items():
+                transform_name = connections["transform"].split(".")[0]
+
+                # Duplicate_transform subd patch geometry.
+                duplicate_transform = cmds.duplicate(transform_name)[0]
+                delete_bin.append(duplicate_transform)
+
+                # Discard the children.
+                shapes = cmds.listRelatives(duplicate_transform, shapes=True)
+                children = cmds.listRelatives(
+                    duplicate_transform, children=True
+                )
+                cmds.delete(set(children) - set(shapes))
```

Comment:
Note: Older versions of maya don't support directly operating on `set` types. We did a bugfix some time ago for some other OP code.

It's Maya 2018 and below by the way, not sure if still relevant?

See PR: https://github.com/ynput/OpenPype/pull/2932

Reply:
Nope gone to 2022 and handed off grooming to Houdini sorry

Reply:
Cool, gunna leave this.